### PR TITLE
Run JS unit tests and e2e tests in parallel on CI

### PR DESCRIPTION
**Pull request by swissspidy**
## Context

While working on parallelizing the Karma tests (see #10186) I wondered whether we could do the same for our Jest tests.

## Summary

Implements sharding for the JS unit tests and the e2e tests by leveraging a custom test sequencer class.

## Relevant Technical Choices

Adds a new private package for the shared test sequencer.

## User-facing changes

N/A

## Testing Instructions

- [x] This is a non-user-facing change and requires no QA

This PR can be tested by following these steps:

1.

### Does this PR have a security-related impact?

No

### Does this PR change what data or activity we track or use?

No

### Does this PR have a legal-related impact?

No

## Checklist

- [ ] This PR addresses an existing issue and I have linked this PR to it in ZenHub
- [x] I have tested this code to the best of my abilities
- [x] I have verified accessibility to the best of my abilities ([docs](https://github.com/google/web-stories-wp/blob/main/docs/accessibility-testing.md))
- [x] I have verified i18n and l10n (translation, right-to-left layout) to the best of my abilities
- [x] This code is covered by automated tests (unit, integration, and/or e2e) to verify it works as intended ([docs](https://github.com/google/web-stories-wp/tree/main/docs#testing))
- [x] I have added documentation where necessary
- [x] I have added a matching `Type: XYZ` label to the PR

---

See #10157 and #10186

### CODE REVIEW

**Comment by spacedmonkey** [comment]
I am only seeing WP 5.5 E2E testing running. 
Screenshot 2022-01-14 at 12 50 36

Is this is the desired behaviour?

**Comment by swissspidy** [comment]
Hmm right, because we're using `include:` for the matrix there I probably need to add all shards manually... Ugh.

**Comment by swissspidy** [comment]
Now they're all there:

Screenshot 2022-01-14 at 14 33 12

**Comment by swissspidy** [comment]
Strangely the `Stories Dashboard › should be able to open the dashboard` e2e test seems to be failing here 🤔

**Comment by spacedmonkey** [comment]
This looks good, but there e2e tests are now failing. I will approve once they are working again.

**Comment by swissspidy** [comment]
All tests are passing now.

File: .github/workflows/tests-e2e.yml [hunk 1]
```diff
@@ -63,24 +63,46 @@ concurrency:
...
+            shard: '2/2'
+        exclude:
+          - browser: 'chrome'
+            wp: 'latest'
+            snapshots: false
+            shard: '1/2'
+          - browser: 'chrome'
+            wp: 'latest'
             snapshots: false
+            shard: '2/2'
```
Comment (samwhale) [inline .github/workflows/tests-e2e.yml]: Just for my own education: why do we have to build each of these out separately?

File: packages/e2e-tests/src/specs/dashboard/adminMenu.js [hunk 2]
```diff
@@ -27,9 +27,9 @@ describe('Admin Menu', () => {
 
     // Initial visit to `/` makes `Dashboard` link current in WP
     await page.hover('#menu-posts-web-story');
-    await await expect(
-      page.$('#menu-posts-web-story .current a')
-    ).resolves.toMatch('Dashboard');
+    await expect(page.$('#menu-posts-web-story .current a')).resolves.toMatch(
```
Comment (samwhale) [inline packages/e2e-tests/src/specs/dashboard/adminMenu.js:30]: 😂 this test file is getting flagged by codecov. I can write a ticket to look into this (unless it's a known problem already)
image

File: packages/jest-parallel-sequencer/README.md [hunk 3]
```diff
@@ -0,0 +1,5 @@
+# Jest Parallel Sequencer
+
+Simple custom [test sequencer](https://jestjs.io/docs/configuration#testsequencer-string) for Jest to allow for sharding on CI since Jest [does not support this out of the box](https://github.com/facebook/jest/issues/11252).
+
+Requires the `SHARD` environment variable to be set.
```
Comment (samwhale) [inline packages/jest-parallel-sequencer/README.md:5]: 🥳 love documentation

File: packages/jest-parallel-sequencer/src/index.js [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+   */
+  sort(tests) {
+    const chunkSize = Math.ceil(tests.length / totalShards);
+    const minIndex = shardIndex * chunkSize;
+    const maxIndex = minIndex + chunkSize;
+
+    return Array.from(tests)
+      .sort((testA, testB) => (testA.path > testB.path ? 1 : -1))
+      .slice(minIndex, maxIndex);
+  }
```
Comment (samwhale) [inline packages/jest-parallel-sequencer/src/index.js:46]: 💭 should we run tests in random orders each time? Could be nice to keep things moving rathe than always running the same tests on the same shards. (Not sure if this already happens)

But if we want known randoms we could randomize the order based off of the current commit on main 🤔

File: tests/js/jest.config.js [hunk 5]
```diff
@@ -41,6 +41,11 @@ export default {
     '**/test/**/*.[jt]s',
     '**/?(*.)test.[jt]s',
   ],
+  // @jest/test-sequencer is the default.
+  testSequencer:
+    'true' === process.env.CI
+      ? '@web-stories-wp/jest-parallel-sequencer'
+      : undefined,
```
Comment (samwhale) [inline tests/js/jest.config.js:48]: 🥳

**Review by samwhale — APPROVED** [review]
This stuff is so cool man. Learning a lot from seeing this stuff!

File: packages/jest-parallel-sequencer/src/index.js (hunk 4 above)
Comment (swissspidy) [reply packages/jest-parallel-sequencer/src/index.js:46]: Good idea!

Running e2e tests in random order could help catch cases like aa46d477aac4461cf4029df4a6f6a7099a37f453 where tests inadvertently depend on the specific order of tests being run. It's why we also run PHP unit tests in random order  in a separate job allowed to fail.

I do think though that this is best handled in a future enhancement, ideally when Jest supports this out of the box (see https://github.com/facebook/jest/issues/4386).

File: packages/e2e-tests/src/specs/dashboard/adminMenu.js (hunk 2 above)
Comment (swissspidy) [reply packages/e2e-tests/src/specs/dashboard/adminMenu.js:30]: Updated the codecov config,  that should fix this. Thanks for flagging!

File: .github/workflows/tests-e2e.yml (hunk 1 above)
Comment (swissspidy) [reply .github/workflows/tests-e2e.yml]: Hmm I can probably simplify this.

Basically we want to run these tests:

* WP 5.5, no snapshots
* WP latest, with snapshots
* WP 5.9 RC, no snapshots, experimental

So this involves using `include` here to add the additional tests.

Thanks to your comment I realized that I could simplify this matrix a bit though.

**Pull request merged**